### PR TITLE
Add brain for numpy core module ``einsumfunc``.

### PR DESCRIPTION
**Pull request by mbyrnepr2**
Refs PyCQA/pylint#5821

## Steps

- [ ] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [ ] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

### CODE REVIEW

File: tests/unittest_brain_numpy_core_einsumfunc.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    HAS_NUMPY = True
+except ImportError:
+    HAS_NUMPY = False
+
+from astroid import builder
+
+
+@unittest.skipUnless(HAS_NUMPY, "This test requires the numpy library.")
+class BrainNumpyCoreFromNumericTest(unittest.TestCase):
```
Comment (DanielNoord) [inline tests/unittest_brain_numpy_core_einsumfunc.py:22]: Could you rewrite these test without using `unittest`? I think we're trying to only add `pytest` style tests but have decided that rewriting old tests is not worth it (atm..)

File: tests/unittest_brain_numpy_core_einsumfunc.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    """
+    Test the numpy core einsumfunc brain module
+    """
```
Comment (DanielNoord) [inline tests/unittest_brain_numpy_core_einsumfunc.py:25]:
```suggestion
    """Test the numpy core einsumfunc brain module."""
```

At some point we should turn on `pydocstringformatter` here as well.

File: ChangeLog [hunk 3]
```diff
@@ -69,6 +69,10 @@ Release date: TBA
 * Fix test for Python ``3.11``. In some instances ``err.__traceback__`` will
   be uninferable now.
 
+* Add brain for numpy core module ``einsumfunc``.
+
+  Refs PyCQA/pylint#5821
```
Comment (DanielNoord) [inline ChangeLog:74]:
```suggestion
  Closes PyCQA/pylint#5821
```

File: astroid/brain/brain_numpy_core_einsumfunc.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+Astroid hooks for numpy.core.einsumfunc module:
+https://github.com/numpy/numpy/blob/main/numpy/core/einsumfunc.py
+"""
+
+from astroid.brain.helpers import register_module_extender
+from astroid.builder import parse
+from astroid.manager import AstroidManager
+
+
+def numpy_core_einsumfunc_transform():
```
Comment (DanielNoord) [inline astroid/brain/brain_numpy_core_einsumfunc.py:15]:
```suggestion
def numpy_core_einsumfunc_transform() -> nodes.Module:
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Not sure why this landed in PR limbo, but I thought I'd start with a review.

More general question: why are we putting al `numpy` brains into separate files? Seems like a bit of a waste, right? Or am I missing something?

**Comment by mbyrnepr2** [comment]
> Not sure why this landed in PR limbo, but I thought I'd start with a review.
> 
> More general question: why are we putting al `numpy` brains into separate files? Seems like a bit of a waste, right? Or am I missing something?

Thanks!

I don't know the story behind this but to my mind it seems more maintainable to have them as separate modules, especially if if they increase in size over time. Perhaps we could create a separate sub-package `astroid/brain/numpy/` to collect them in one place?

Or perhaps your point is more related to performance @DanielNoord?

**Comment by DanielNoord** [comment]
> I don't know the story behind this but to my mind it seems more maintainable to have them as separate modules, especially if they increase in size over time. 

Yeah I think a refactor might be beneficial as I'm not sure this is actually more maintainable. For example, the brain you added could have been added to https://github.com/PyCQA/astroid/blob/main/astroid/brain/brain_numpy_core_fromnumeric.py.
That would have removed 17 unnecessary lines from this diff in the brain file alone.

Same probably goes for test files.

**Comment by mbyrnepr2** [comment]
Hi @DanielNoord I think you have good ideas. As a way forward may I suggest you create a new issue to capture the ideas behind your refactoring suggestion? I'm sure there are knowledgable folks who are aware of the reasoning behind the existing structure who can provide input & help move that forward.

The current issue can move forward independently & according to the existing structure; if the refactoring issue comes to fruition then it wouldn't be much extra work to incorporate einsumfunc brain into the new system.

What do you think?

**Comment by DanielNoord** [comment]
I'm fine with creating a new issue, but I'm not sure if there is anybody around anymore that knows the original rationale.

Why don't approach this from what we would want ourselves. If you think keeping these separate is better that's fine with me, but if you think it's better to merge them then let's do so now!

**Comment by mbyrnepr2** [comment]
Mainly I think for now I'd like to follow the existing structure in order to stick to the scope of fixing the Pylint issue; hopefully I have time to finish that much this week.
Regarding the refactor itself - I'm in two minds because I think the structure now is helpful with the 1:1 of brain to numpy module mapping; but perhaps I am missing something.

File: tests/unittest_brain_numpy_core_einsumfunc.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@pytest.mark.skipif(not HAS_NUMPY, reason="This test requires the numpy library.")
+@pytest.mark.parametrize(
+    "method, args",
+    [
+        ("einsum", "ii, np.arange(25).reshape(5, 5)"),
+    ],
+)
+def test_numpy_function_calls_inferred_as_ndarray(method, args):
```
Comment (DanielNoord) [inline tests/unittest_brain_numpy_core_einsumfunc.py:37]:
```suggestion
def test_numpy_function_calls_inferred_as_ndarray(method: str, args: str) -> None:
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: tests/unittest_brain_numpy_core_einsumfunc.py (hunk 5 above)
Comment (mbyrnepr2) [reply tests/unittest_brain_numpy_core_einsumfunc.py]: This isn't relevant if we remove the parametrize decorator.

**Review by DanielNoord — APPROVED** [review]

**Comment by DanielNoord** [comment]
Pushed some final changes. Thanks @mbyrnepr2 😄

**Pull request merged**